proxy/client.go

```diff
@@ -0,0 +1,226 @@
+package proxy
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/rollkit/go-execution"
+	proxy_json_rpc "github.com/rollkit/go-execution/proxy/jsonrpc"
+	rollkitTypes "github.com/rollkit/rollkit/types"
+)
+
+type PayloadStatus string
+
+const (
+	PayloadStatusValid   PayloadStatus = "VALID"
+	PayloadStatusInvalid PayloadStatus = "INVALID"
+	PayloadStatusSyncing PayloadStatus = "SYNCING"
+)
+
+var (
+	ErrNilPayloadStatus     = errors.New("nil payload status")
+	ErrInvalidPayloadStatus = errors.New("invalid payload status")
+)
+
+// Ensure ProxyClient implements Execute interface
+var _ execution.Execute = (*ProxyClient)(nil)
+
+// ProxyClient implements the Execute interface in go-execution
+type ProxyClient struct {
+	client       *proxy_json_rpc.Client
+	engineClient *rpc.Client // engine api
+	ethClient    *ethclient.Client
+	genesisHash  common.Hash
+	feeRecipient common.Address
+}
+
+func NewClient(config *proxy_json_rpc.Config, ethURL, engineURL string, genesisHash common.Hash, feeRecipient common.Address) (*ProxyClient, error) {
+	client := proxy_json_rpc.NewClient()
+	client.SetConfig(config)
+
+	ethClient, err := ethclient.Dial(ethURL)
+	if err != nil {
+		return nil, err
+	}
+
+	engineClient, err := rpc.Dial(engineURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ProxyClient{
+		client:       client,
+		engineClient: engineClient,
+		ethClient:    ethClient,
+		genesisHash:  genesisHash,
+		feeRecipient: feeRecipient,
+	}, nil
+}
+
+// Start starts the proxy client
+func (p *ProxyClient) Start(url string) error {
+	return p.client.Start(url)
+}
+
+// Stop stops the proxy client
+func (p *ProxyClient) Stop() {
+	p.client.Stop()
+}
+
+// Implement the Execute interface
+func (p *ProxyClient) InitChain(genesisTime time.Time, initialHeight uint64, chainID string) (rollkitTypes.Hash, uint64, error) {
+	ctx := context.Background()
```

Comment:
Ah, this reminded me to create https://github.com/rollkit/go-execution/issues/25 ;) Thanks!